Improve GDExtension user-friendliness by being more explicit on when return value should be passed initialized

touilleMan opened the pull request:
fix #34264

ptrcall always overwrite the return value buffer with it own return value
data. Hence a non initialized buffer should be provided to ptrcall.

I've tested this with code like this:
```c
godot_string __ret;
memset(&__ret, 0xff, sizeof(__ret));
gdapi10.godot_method_bind_ptrcall(
    __methbind__Object__get_class,
    my_obj_ptr,
    NULL,
    &__ret
)
```
without the fix this endup in a segfault, with this fix it works fine

I guess we should be double careful before merging this given ptrcall is a strange beast and my knowledge is limited on it ;-)

touilleMan commented:
Issue #35609 is also related to this

Considering this code (see https://github.com/touilleMan/godot-python/commit/d0f22451867e2fd4ce7dd5a196e2ca126c2aa804):
```C
godot_object *__ret;  // C doesn't automatically do zero initialization
gdapi10.godot_method_bind_ptrcall(
    __methbind__Object__get_class,  // Object.get_class returns a Reference object
    my_obj_ptr,
    NULL,
    &__ret  // Reference.unreference will be called on __ret as part of it destructor
            // hence causing a segfault given __ret points on nothing
)
```

touilleMan commented:
I've rebased this branch on the master, should be ready to merge ;-)

touilleMan commented:
@vnen can you have a look at this ? 😃

neikeq commented:
Not entirely sure what this does. How would this affect engine code (not GDNative) like the following?:

``` C++
String ret;
method->ptrcall(ptr, nullptr, &ret);

Ref<Reference> ret;
method->ptrcall(ptr, nullptr, &ret);
```

This is what the glue for C# uses.

touilleMan commented:
@neikeq thanks you for your input, this is a very interesting

> How would this affect engine code (not GDNative) like the following?:

The initial `String` object would be overwritten without proper call to it destructor.

This is "fine" in your example given `String` constructor doesn't do any heap allocation:
https://github.com/godotengine/godot/blob/0ee744ff5f3cf045a2699aa4111de3200b9d70be/core/string/ustring.h#L438
and neither does it CowData field:
https://github.com/godotengine/godot/blob/0ee744ff5f3cf045a2699aa4111de3200b9d70be/core/templates/cowdata.h#L63

It feels however very fragile:
- if `String` initialization changes it could lead to hard to track memory leaks
- it seems very strange that `String()` doesn't cause memory leak but `String("foo")` does... (this would bite people trying to save some space in there stack by re-using a string variable in their code)
- ptrcall doesn't only returns string objects, there could be other objects

So I would say there different expectation between the C and the C++ API here:
in C, developer expects to do:
```c
godot_string ret;
gdapi10.godot_method_bind_ptrcall(
    __methbind__Object__get_class,
    my_obj_ptr,
    NULL,
    &ret
)
[...]
godot_string_destroy(&ret)
```
in C++, developer expects to do:
```c++
String ret;
method->ptrcall(ptr, nullptr, &ret);
[...]
```

Currently the C API forces you to do the C++ way (which leads to segfault), this PR forces the C++ API to do the C way (which leads to memory leaks as you pointed out ^^)

The solutions I see:
- Make `godot_method_bind_ptrcall` behave differently from `method->ptrcall`. I guess this would require introducing a `method->ptrcall_with_uninitialized_ret`
- Add runtime check (disabled for release builds ?) to ensure the provided ret object is correct (I guess the only check we can do is "is the pointed data null initialized ?", this might be a problem for object other than strings)
- Choose one way of doing and document it well (but ptrcall&gnative documentation are virtually existent, and it very easy to miss a comment lost in the codebase anyway...)

neikeq commented:
>- ptrcall doesn't only returns string objects, there could be other objects

Can confirm this is already the case with `Array`, which does allocate in the default constructor:

https://github.com/godotengine/godot/blob/dfe4feb1884ba058f2d8938ed13221949d6cc483/core/variant/array.cpp#L571-L574

So this PR would cause leaks with the C# glue:

https://gist.github.com/neikeq/857a621f2878135ce1e2a702cd0a2bb9#file-mono_glue-gen-cpp-L119

neikeq commented:
I agree with the change and I don't have a problem updating the C# glue. I'm committing some changes soon which will make this change easier. However, this is a breaking change so we need opinion from the @godotengine/gdnative team. There's also the concern you mention about someone passing a constructed object like `Array` which would leak if not destructed; but ptrcall was always a very strict internal feature and you should understand it well before using it.

aaronfranke commented:
@touilleMan GDNative has been replaced with GDExtension. Does this problem still exist on the current master? If so, this PR needs to be rebased on the latest master branch.

vnen commented:
I don't think this should be done. AFAIK you're expected to pass a constructed value to ptrcall. If the problem is the C API then I think it should be done only for the C API.

touilleMan commented:
This PR is totally outdated, but the issue still remains 😄 (tl;dr: well maybe there is only a documentation issue after all)

So here is an attempt to provide an up-to-date view on the issue (hence ping @godotengine/gdextension !):

## 1 - How things work inside Godot

Godot is a C++ codebase, so all the builtins classes (e.g. `Variant`, `Vector2`, `String` etc.) get their constructor automatically called when a value of their type is created.

So it's fine to have method that get output parameter as reference (i.e. `r_ret` here):
https://github.com/godotengine/godot/blob/57bdddce02370af3d13e6edc583afdc17264147d/core/variant/variant.h#L538

and use them this way:
https://github.com/godotengine/godot/blob/57bdddce02370af3d13e6edc583afdc17264147d/core/variant/variant.cpp#L864-L866

And finally in `evaluate` we have typically something like
https://github.com/godotengine/godot/blob/57bdddce02370af3d13e6edc583afdc17264147d/core/variant/variant_op.cpp#L1027

So here `r` is initialized once by `Variant`'s default constructor, then passed as reference to `evaluate` which in turn creates it own `Variant` object (hence calling `Variant`'s constructor) that overwrite `r` (hence `Variant`'s destructor is called on `r` just before the overwrite occurs)

(one could argue that `r` is initialized twice)

## 2 - How things works with GDExtension

On the other hand, GDExtension is a C API, so the `evaluate` is exposed with a different signature:
https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/extension/gdnative_interface.h#L414

Being a C API, the user would expect to be able to do:
```c
gd_variant_t return_value;  // Let's pretend gd_variant_t is defined as a C opaque struct of sizeof(Variant)
GDNativeBool valid;
godot_api->variant_evaluate(p_op, p_a, p_b, &return_value, &valid)
```

This will work fine as long as the compiler does zero initialization, which [is not guaranteed in C](https://stackoverflow.com/a/1597491) ([unlike for C++](https://en.cppreference.com/w/cpp/language/zero_initialization)).
So basically your code seems to work until you change your compiler or try to make it work on another OS then you get a nasty segfault :'(

This are getting worst when doing ptrcall on Godot classe method given some of them returns a `String`
https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/string/ustring.h#L182-L184

There is not destructor for `String`, meaning the default one is used which will call `_cowdata`'s destructor...
https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/templates/cowdata.h#L413-L415

...which will lead to a segfault :'(

`String` is not alone on this:

| type                | has destructor   | destructor supports zero initialized object |
|---------------------|------------------|---------------------------------------------|
| NIL                 |  no ✅           | n/a |
| BOOL                |  no ✅           | n/a |
| INT                 |  no ✅           | n/a |
| FLOAT               |  no ✅           | n/a |
| VECTOR2             |  no ✅           | n/a |
| VECTOR2I            |  no ✅           | n/a |
| RECT2               |  no ✅           | n/a |
| RECT2I              |  no ✅           | n/a |
| VECTOR3             |  no ✅           | n/a |
| VECTOR3I            |  no ✅           | n/a |
| TRANSFORM2D         |  no ✅           | n/a |
| VECTOR4             |  no ✅           | n/a |
| VECTOR4I            |  no ✅           | n/a |
| PLANE               |  no ✅           | n/a |
| QUATERNION          |  no ✅           | n/a |
| AABB                |  no ✅           | n/a |
| BASIS               |  no ✅           | n/a |
| TRANSFORM           |  no ✅           | n/a |
| PROJECTION          |  no ✅           | n/a |
| COLOR               |  no ✅           | n/a |
| RID                 |  no ✅           | n/a |
| OBJECT (i.e. pointer)|  no ✅           | n/a |
| VARIANT             |  yes ⚠️          | ✅ [yes](https://github.com/godotengine/godot/blob/57bdddce02370af3d13e6edc583afdc17264147d/core/variant/variant.h#L301) |
| STRING              |  yes ⚠️          | ✅ yes (calls [`CowData`'s destructor](https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/templates/cowdata.h#L197-L199)) |
| STRING_NAME         |  yes ⚠️          | ✅ [yes](https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/string/string_name.h#L180) |
| NODE_PATH           |  yes ⚠️          | ✅ [yes](https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/string/node_path.cpp#L96) |
| CALLABLE            |  yes ⚠️          | ✅ [yes](https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/variant/callable.cpp#L310) |
| SIGNAL              |  yes ⚠️          | ✅ yes (calls `StringName`'s destructor]) |
| DICTIONARY          |  yes ⚠️          | ⚠️ yes (but [logs an error](https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/variant/dictionary.cpp#L290)) |
| ARRAY               |  yes ⚠️          | ✅ yes |
| PACKED_BYTE_ARRAY   |  yes ⚠️          | ✅ yes (calls [`CowData`'s destructor](https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/templates/cowdata.h#L197-L199)) |
| PACKED_INT32_ARRAY  |  yes ⚠️          | ✅ yes (calls [`CowData`'s destructor](https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/templates/cowdata.h#L197-L199)) |
| PACKED_INT64_ARRAY  |  yes ⚠️          | ✅ yes (calls [`CowData`'s destructor](https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/templates/cowdata.h#L197-L199)) |
| PACKED_FLOAT32_ARRAY|  yes ⚠️          | ✅ yes (calls [`CowData`'s destructor](https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/templates/cowdata.h#L197-L199)) |
| PACKED_FLOAT64_ARRAY|  yes ⚠️          | ✅ yes (calls [`CowData`'s destructor](https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/templates/cowdata.h#L197-L199)) |
| PACKED_STRING_ARRAY |  yes ⚠️          | ✅ yes (calls [`CowData`'s destructor](https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/templates/cowdata.h#L197-L199)) |
| PACKED_VECTOR2_ARRAY|  yes ⚠️          | ✅ yes (calls [`CowData`'s destructor](https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/templates/cowdata.h#L197-L199)) |
| PACKED_VECTOR3_ARRAY|  yes ⚠️          | ✅ yes (calls [`CowData`'s destructor](https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/templates/cowdata.h#L197-L199)) |
| PACKED_COLOR_ARRAY  |  yes ⚠️          | ✅ yes (calls [`CowData`'s destructor](https://github.com/godotengine/godot/blob/4ab3fdcda07bf700b33eb7c85f53826b2464c02f/core/templates/cowdata.h#L197-L199)) |

So to summarize, forcing zero-initialization seems a good rule of thumb to avoid be safe when using GDExtension API, however:

1) Dictionary still creates an error logs (but we could iron this out easily ^^)
2) It is an unexpected requirement, so we must add it in the documentation, then people won't read it and will get unexpected and hard-to-track issues :'(

## 3 - Zero-initialization on C is hard !

On top of that doing zero initialization is [suprizingly error prone in C](https://thephd.dev/ever-closer-c23-improvements#consistent-warningless-and-intuitive-initialization-with--):
The best way to do it is

```c
gd_variant_t a;
memset(&a, 0, sizeof(a));
```

However I've just written a buggy code because it should be written [`sizeof a` or `sizeof(gd_variant_t)`](https://twitter.com/thingskatedid/status/1367883687237279745) (cf. [sizeof syntax](https://learn.microsoft.com/en-us/cpp/c-language/sizeof-operator-c?view=msvc-170#syntax)). Now good luck to debug this issue that occurs only some time on some Godot methods ;-)

Also C++ programmer are often eager to use `gd_variant_t a = { 0 };` as zero-init, but in C this [only initializes the first element](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:7,endLineNumber:3,positionColumn:7,positionLineNumber:3,selectionStartColumn:7,selectionStartLineNumber:3,startColumn:7,startLineNumber:3),source:'struct+foo++%7B%0A+++int+a%3B%0A+++int+b%3B%0A%7D%3B%0A%0Avoid+blah(struct+foo+*p)%3B%0A%0Avoid+foo(void)+%7B%0A++++struct+foo+c+%3D+%7B+0+%7D%3B%0A++++blah(%26c)%3B%0A%7D'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cg101,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'1',libraryCode:'1',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-O2',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'x86-64+gcc+10.1+(C,+Editor+%231,+Compiler+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+10.1+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) !

Of course it depends on how `gd_variant_t` is defined... which is left up to the user of GDExtension.

The most obvious way of defining an opaque structure would be to make it include a singe char buffer field:
```c
struct gd_variant_t  {
   char opaque[3];
};
```

So now given we have only one attribute in the structure we can do this fancy `gd_variant_t a = { 0 };`...
...and we may get a segfault because the structure may [contains padding bytes whose value is undefined](https://interrupt.memfault.com/blog/c-struct-padding-initialization#the-current-state).

## 4 - Bonus: Correctly implementing return value as parameter is tricky !

### 1 - In Godot

Considering how `gdnative_variant_call` is implemented:
https://github.com/godotengine/godot/blob/57bdddce02370af3d13e6edc583afdc17264147d/core/extension/gdnative_interface.cpp#L82-L89

Line 89 we use `memnew_placement` that doesn't call destructor on `r_return` before overwritting it content.
This works fine until the caller starts reusing it objects:

```c
gd_variant_t ret;
memset(&ret, 0, sizeof(gd_variant_t));
for (int i = 0; i++; i < 10) {
    // No need to put memset&destructor in the loop given `gdnative_variant_call` takes care of that !
    godot_api->gdnative_variant_call(&self, &method, NULL, 0, &ret, NULL);
}
godot_api->variant_destroy(&ret);
```

And now we've leaked data :'(

### 2 - In native extension

GDExtension allows user to register they own native extension, which should provide call&ptrcall function pointers for each methods:

```c
typedef void (*GDNativeExtensionClassMethodCall)();
typedef void (*GDNativeExtensionClassMethodPtrCall)();

static void my_extension_do_something_call(void *method_userdata, GDExtensionClassInstancePtr p_instance, const GDNativeVariantPtr *p_args, const GDNativeInt p_argument_count, GDNativeVariantPtr r_return, GDNativeCallError *r_error) {
    GDNativeCallError error
    // Oups ! we forget to call `godot_api->variant_destroy(r_return)`, now we leak memory !
    godot_api->variant_constructor(GDNATIVE_VARIANT_TYPE_DICTIONARY, r_return, NULL, 0, &error);
    // Continue to populate the dictionary
}
static void my_extension_do_something_ptrcall_v1(void *method_userdata, GDExtensionClassInstancePtr p_instance, const GDNativeTypePtr *p_args, GDNativeTypePtr r_ret) {
    // r_ret is already a valid Dictionary given it has been initialized by the caller, let's use it right away...
    // ...but this is only true when the caller is Godot, if the caller is another native extension r_ret might just be
    // zero-initialized and we will get a segfault when using it :(
}
static void my_extension_do_something_ptrcall_v2(void *method_userdata, GDExtensionClassInstancePtr p_instance, const GDNativeTypePtr *p_args, GDNativeTypePtr r_ret) {
    // Oups ! we forget to call dictionary destructor first ! Dictionary does heap alloc on init so we most likely leak memory !
    godot_dictionary_constructor(r_ret);
    // Continue to populate the dictionary
}
```

So the rule of thumb for native extension developer is "always start by calling destructor on the return value" which seems
very easy to forget.
And if the extension developer does forget, this has not consequence most of the time. Until the native extension *user*
starts reusing objects in GDExtension. This looks like a recipe for disaster to me 😄 

## 5 - What can we do about this ?

So to summarize:

- not initializing return value leads to segfault
- initialializing return value with constructor leads to memory leak given its complicated to know if the destructor is called

I see three possibilities to fix this, by order of simplicity:

### 1 - Always use constructor policy

The rule of thumb becomes:

> For Variant and builtins that are marked `has_destructor: true`
>
> 1) always call the constructor on return value before passing it in a call to GDExtension API
> 2) always call the destructor on return value before using it in native extension

This is the cheapest solution given we only have to add huge stickers everywhere to tell about the rule of thumb (typically in GDExtension documentation when it will be available and in `gdnative_interface.h`)

Of course the downside is that it leaves the burden on GDExtension users and native extension developer.
However this might be considered ok given GDExtension is a low-level API and actual game developers are expected to use wrappers as Godot-CPP/Godot-Rust/Godot-Python etc.

### 2 - Don't do allocation on init

This is a variation of the previous, the idea is to guarantee that all builtins don't do allocation on init. This way not calling the destructor is ok as long as the object has not been used yet.

This has the added benefit of allowing to do initialization with zero-initialization instead of calling the constructor (but as we saw, this is error prone so I'm not sure how much of a benefit this is 😄 )

The main benefit is it remove the necessity for native extension developer to call the destructor on return value.

However removing allocation on init [might cause issue with how array/dictionnary behave](https://github.com/godotengine/godot/issues/35779#issuecomment-581299134)...

### 3 - Stick with the C convension

This is the approach I originally implemented in this PR. We would have to make sure the GDExtension API only expose functions
that consider the pointer provided as return value contains unitialized memory.

The trick is, as @vnen said, we have to make sure this C API stays isolated and doesn't leak in the rest of the codebase.

This is obvious enough for most of the GDExtension API given it return value type is a `void *` or `GDNativeTypePtr` where Godot internal works with `Variant *`/`Vector2 *`/`String *` etc.
The tricky part is for the functions pointers returned (`GDNativePtrOperatorEvaluator`, `GDNativePtrBuiltInMethod` and `GDNativePtrUtilityFunction`) that are shared between C and C++.
So we would need two flavours of ptrcall (ptrcall and cptrcall ?), the C one would do default initialization on the return value then call the C++ one (except for `NativeExtensionMethodBind` that would implement the C++ version by calling into the C version that is provided by the native extension)

The drawback is the increased complexity of the code base and the size of the final binary to have this second flavor of ptrcall :/

### 4 - Modify GDExtension API to never pass return value by parameter

The issue here comes from the fact we pass the return value as parameter. If we were returning the return value we wouldn't have this issue in the first place !
Of course this means modifying a lot of code in the Godot codebase, and I'm not sure this is even possible (I guess the reason for passing the return value as parameter is it allow to have opaque return value of arbitrary size).
Anyway this "solution" is me being curious of an explanation ;-)

## 6 - Conclusion

It's 2am, I've spent ~8h on this topic and I end up concluding GDExtension has already implemented the "least worst" solution and there is nothing to do here, talking of wasted time 😭 

More seriously I hope this research work will help when writing GDExtension documentation to provide the right guidelines information. I guess we can start by adding a comment about that in `gdnative_interface.h`, what do you think ?

touilleMan commented:
## 1 - More graph

After a good night of sleep, here is an exaustive view of all the places we use return value as pointer in GDExtension:

| name                                              | part of `GDNativeInterface` / or function pointer ? | return value type  | GDExtension specific | destructor first called on return value ? |
|---------------------------------------------------|---------------------------------------------------|--------------------|----------------------|-------------------------------------------|
| GDNativeVariantFromTypeConstructorFunc            | function ptr                                      | GDNativeVariantPtr | yes                  | ⚠️ [yes](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/variant/variant_internal.h#L1525)
| GDNativeTypeFromVariantConstructorFunc            | function ptr                                      | GDNativeTypePtr    | yes                  | ⚠️ [yes](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/variant/variant_internal.h#L1531)
| GDNativePtrOperatorEvaluator                      | function ptr                                      | GDNativeTypePtr    | no                   | ⚠️ [yes](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/variant/variant_op.h#L53) (by using [PtrToArg<R>::encode](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/variant/method_ptrcall.h#L49))
| GDNativePtrBuiltInMethod                          | function ptr                                      | GDNativeTypePtr    | no                   | ⚠️ yes (builtins only implemented in C++ inside Godot)
| GDNativePtrGetter                                 | function ptr                                      | GDNativeTypePtr    | no                   | ⚠️ [yes](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/variant/variant_setget.h#L54) (by using [PtrToArg<R>::encode](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/variant/method_ptrcall.h#L49))
| GDNativePtrIndexedGetter                          | function ptr                                      | GDNativeTypePtr    | no                   | ⚠️ [yes](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/variant/variant_setget.cpp#L350) (by using [PtrToArg<R>::encode](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/variant/method_ptrcall.h#L49))
| GDNativePtrKeyedGetter                            | function ptr                                      | GDNativeTypePtr    | no                   | ⚠️ [yes](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/variant/variant_setget.cpp#L985) (by using [PtrToArg<R>::encode](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/variant/method_ptrcall.h#L49))
| GDNativePtrUtilityFunction                        | function ptr                                      | GDNativeTypePtr    | no                   | yes (all utility functions are defined in C++ inside Godot)
| GDNativeExtensionClassGet                         | function ptr                                      | GDNativeVariantPtr | yes                  | ⚠️ implemented by native extension dev, so maybe
| GDNativeExtensionClassPropertyGetRevert           | function ptr                                      | GDNativeVariantPtr | yes                  | ⚠️ implemented by native extension dev, so maybe
| GDNativeExtensionClassCallVirtual                 | function ptr                                      | GDNativeTypePtr    | yes                  | ⚠️ implemented by native extension dev, so maybe
| GDNativeExtensionClassMethodCall                  | function ptr                                      | GDNativeVariantPtr | yes                  | ⚠️ implemented by native extension dev, so maybe
| GDNativeExtensionClassMethodPtrCall               | function ptr                                      | GDNativeTypePtr    | yes                  | ⚠️ implemented by native extension dev, so maybe
| GDNativeExtensionScriptInstanceGet                | function ptr                                      | GDNativeVariantPtr | yes                  | ⚠️ implemented by native extension dev, so maybe
| GDNativeExtensionScriptInstancePropertyGetRevert  | function ptr                                      | GDNativeVariantPtr | yes                  | ⚠️ implemented by native extension dev, so maybe
| GDNativeExtensionScriptInstanceCall               | function ptr                                      | GDNativeVariantPtr | yes                  | ⚠️ implemented by native extension dev, so maybe
| variant_new_copy                                  | gdnative interface                                | GDNativeVariantPtr | yes                  | [no](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L71)
| variant_new_nil                                   | gdnative interface                                | GDNativeVariantPtr | yes                  | [no](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L74)
| variant_call                                      | gdnative interface                                | GDNativeVariantPtr | yes                  | [no](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L89)
| variant_call_static                               | gdnative interface                                | GDNativeVariantPtr | yes                  | [no](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L105)
| variant_evaluate                                  | gdnative interface                                | GDNativeVariantPtr | yes                  | ⚠️ [yes](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L118)
| variant_get                                       | gdnative interface                                | GDNativeVariantPtr | yes                  | [no](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L170)
| variant_get_named                                 | gdnative interface                                | GDNativeVariantPtr | yes                  | [no](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L179)
| variant_get_keyed                                 | gdnative interface                                | GDNativeVariantPtr | yes                  | [no](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L188)
| variant_get_indexed                               | gdnative interface                                | GDNativeVariantPtr | yes                  | [no](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L197)
| variant_iter_init                                 | gdnative interface                                | GDNativeVariantPtr | yes                  | ⚠️ [yes](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L205)
| variant_iter_next                                 | gdnative interface                                | GDNativeVariantPtr | yes                  | [used as both input&output param](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L218)
| variant_iter_get                                  | gdnative interface                                | GDNativeVariantPtr | yes                  | [no](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L228)
| variant_duplicate                                 | gdnative interface                                | GDNativeVariantPtr | yes                  | [no](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L256)
| variant_stringify                                 | gdnative interface                                | GDNativeStringPtr  | yes                  | [no](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L261)
| variant_get_type_name                             | gdnative interface                                | GDNativeStringPtr  | yes                  | [no](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L290)
| variant_get_constant_value                        | gdnative interface                                | GDNativeVariantPtr | yes                  | [no](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L522)
| string_new_with_latin1_chars                      | gdnative interface                                | GDNativeStringPtr  | yes                  | no
| string_new_with_utf8_chars                        | gdnative interface                                | GDNativeStringPtr  | yes                  | no
| string_new_with_utf16_chars                       | gdnative interface                                | GDNativeStringPtr  | yes                  | no
| string_new_with_utf32_chars                       | gdnative interface                                | GDNativeStringPtr  | yes                  | no
| string_new_with_wide_chars                        | gdnative interface                                | GDNativeStringPtr  | yes                  | no
| string_new_with_latin1_chars_and_len              | gdnative interface                                | GDNativeStringPtr  | yes                  | no
| string_new_with_utf8_chars_and_len                | gdnative interface                                | GDNativeStringPtr  | yes                  | no
| string_new_with_utf16_chars_and_len               | gdnative interface                                | GDNativeStringPtr  | yes                  | no
| string_new_with_utf32_chars_and_len               | gdnative interface                                | GDNativeStringPtr  | yes                  | no
| string_new_with_wide_chars_and_len                | gdnative interface                                | GDNativeStringPtr  | yes                  | no
| object_method_bind_call                           | gdnative interface                                | GDNativeVariantPtr | yes                  | no
| object_method_bind_ptrcall                        | gdnative interface                                | GDNativeTypePtr    | ⚠️ no                 | ⚠️ yes for Godot methods, maybe for native extension methods
| get_library_path                                  | gdnative interface                                | GDNativeStringPtr  | yes                  | ⚠️ [yes](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/native_extension.cpp#L262)
| GDNativePtrConstructor                            | function ptr                                      | GDNativeTypePtr    | ⚠️ no                 | [no](https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/variant/variant_construct.h#L52)

Column "part of `GDNativeInterface` / or function pointer ?" indicate if this component is accessed from the structure that Godot passed to an extension
(e.g. `godot_api->variant_new_nil(&my_var)`) vs the function pointer that are returend by this api and must be called later on (e.g. `GDNativePtrConstructor *c = godot_api->variant_get_ptr_constructor(GDNATIVE_VARIANT_TYPE_RID, 0); c(&my_rid)`).

The "destructor first called on return value" column should be understood as "can this segfault if I pass an uninitialized return value ?".
From a different point of view, this can also be understood as "if I pass a return value that have done heap allocation, will I get a memory leak ?".

The "GDExtenion specific" column indicate the functions that are only implemented for GDExtension (and hence where we are free to change it behavior without impacting the rest of the Godot internals).

A couple of remarks:

1) All functions that are not GDExtension specific expect to work with initialized return value (well except `GDNativePtrConstructor` given it *is* the constructor). This cannot be fixed without duplicating the API between C++ and C style.
2) All methods in `GDNativeInterface` except `object_method_bind_ptrcall` are specific to GDExtension. Hence it would be good to make them all behave consistently (typically having them all expecting uninitialized return value).
3) `object_method_bind_ptrcall` is a special case, it expect to work with initialized return
4) All the native extension function pointer are implemented by 3rd party so there is no guarantee whatsoever here.

## The updated conclusion

Provided my study is right, it is currently very hard to tell when return value should be initialized and when it should not.
In practice I guess it's not *that* bad as long as user follows what C++ does by always initializing it return value (and so never reuse an existing value).

However I think we can make this API much more consistent by doing some small changes:

- Correct `variant_evaluate`/`variant_iter_init`/`variant_iter_get`/`get_library_path` to work with unitialize return value. This would make `GDNativeInterface` much more coherent given it would make all function except `object_method_bind_ptrcall` work the same way.
- Add a coment to explain than `extension_api.json`'s `has_destructor` determines if the type must be initialized to prevent segfault, and on top of that than zero-initialization should not be considered safe and only the actual constructor can be used when `has_destructor: true`.
- Introduce `GDNativeUninitializedVariantPtr`/`GDNativeUninitializedTypePtr` types. This way `GDNativeVariantPtr`/`GDNativeTypePtr` would represent a Variant or builtin type that is guaranteed to be initialized by the caller where `GDNativeUninitializedVariantPtr`/`GDNativeUninitializedTypePtr` would represent an uninitialized buffer of the size of a Variant or builtin type.

This way we would have a much more explicit and consistent API ;-)

## The update PR \o/

I've updated this PR to implemented the improvement I was talking about.
@godotengine/gdextension What do you think about that ?

reduz commented:
I am not sure if I understand this. Values should be _always_ initialized and uninitialized using GDExtension.

reduz commented:
Would it help to have information in the API JSON dump for each built-in type whether it needs to be constructed/deconstructed before/after use?

touilleMan commented:
> I am not sure if I understand this

tl;dr: it's complicated for user of the GDExtension API to know when the return value should be initialized and when it should not. This PR introduce a type to make this explicit. 😃 

> Values should be always initialized and uninitialized using GDExtension, even the returned ones.

It's not currently the case (see the array in my previous post), for instance:

https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L98-L105

vs: 
https://github.com/godotengine/godot/blob/e5594c26b14e75d8b75d1f697cf2bfbd6254a50c/core/extension/gdnative_interface.cpp#L114-L121

So caller must initialize the return value when using `gdnative_variant_evaluate` but should not with `gdnative_variant_call_static` (calling `gdnative_variant_call_static` with a null variant is no big deal, but reusing a variant that contains an array leads to a memory leak...)

My point here is to find a way to make this more consistent and more obvious for the GDExtension user 😃 

> Would it help to have information in the API JSON dump for each built-in type whether it needs to be constructed/deconstructed before/after use?

I don't think so:
- We already have the `has_destructor` flag that is good enough to know if it is safe to skip initializing a type[^1].
- this "return value param must be initialized/uninitialized" is for the functions defined in `gdnative_interface.h` (for instance `object_method_bind_ptrcall` should always have the return value initialized given it makes us jump into Godot's C++ function, on the other hand `gdnative_variant_new_nil` is a constructor implemented as a function only used for gdextension and hence the return value should be uninitialized)

[^1]: In theory `has_destructor` is not strictly equivalent to `skipping_initialization_is_ok` (for instance the type could contain a pointers on a global value: no need to call a destructor, but not calling the constructor would lead to segfault when using this type). But this is not an issue in Godot given all the types without destructor are only simple structures made of integers.

touilleMan commented:
rebased

touilleMan commented:
@YuriSizov I've seen you've marked changed the milestone for this PR to 4.1

I'm okay with that (I've already talked with @akien-mga during FOSDEM :belgium:  :beers: about the fact GDExtension in Godot 4.0 will be considered beta and so breaking changes are allowed)

However I'd just want to point out  the actual breaking changes this PR contains: the GDExtension code written before this PR should still work but will leak a small amount of memory when calling certain GDExtension API function (e.g. `gdextension_string_new_with_utf8_chars`, or when doing variant to/from type conversion)

This should be easy to fix for GDExtension code writer (especially given documenting which return value passed as parameter should be pre-initialized or not is precisily the point of this PR :smiley: )

YuriSizov commented:
Even if it contains breaking changes, at this point we don't include any breaking changes in 4.0. GDExt is expected to be expanded upon in 4.1, so that's okay.

dsnopek commented:
Discussed at the GDExtension meeting and we think it makes sense.

Just a rebase needed and fixes for the CI errors

Also, @Bromeon will check with the Rust binding and see if this causes any issues

touilleMan committed:
Improve user-friendliness of "return value as function param" pattern in GDExtension API

This commit introduce separate types (e.g. GDNativeStringPtr vs GDNativeUninitializedStringPtr)
depending on if the pointed data is already initialized (C++ style where constructor is alway
called when create a variable even if it is to be passed as return value) or not (C style).
On top of that, small changes has been made to `GDNativeInterface` so that it methods are
consistent on using uninitialized return value.

touilleMan commented:
@dsnopek @Bromeon I've updated this PR, it's should be ready for merging \o/

dsnopek reviewed (approved):
Thanks, and approved! (As that's what we agreed at the GDExtension meeting on 2023-04-24)

While I was looking at it again, I noticed that the wording on the comment explaining the new typedefs could be improved, so if you have time to update that, it'd be great. But we could always fix it up afterwards too, I don't necessarily want that to hold up committing this just because of the comment.

@Bromeon It would still be great if could test this with the Rust bindings, if you have time!

And, if I've understood correctly, this will lead to a small number of situations where godot-cpp could leak memory until its updated? Is there there already a follow-up PR to address that?

Bromeon commented:
[I ran our CI against this commit](https://github.com/godot-rust/gdext/actions/runs/4977544805/jobs/8906731599), however there seem to be changes in `gdextension_interface.h` that are breaking from a Rust perspective:
```diff
  typedef void *GDExtensionVariantPtr;
  typedef const void *GDExtensionConstVariantPtr;
+ typedef void *GDExtensionUninitializedVariantPtr;
  typedef void *GDExtensionStringNamePtr;
  typedef const void *GDExtensionConstStringNamePtr;
+ typedef void *GDExtensionUninitializedStringNamePtr;
  typedef void *GDExtensionStringPtr;
  typedef const void *GDExtensionConstStringPtr;
+ typedef void *GDExtensionUninitializedStringPtr;
  typedef void *GDExtensionObjectPtr;
  typedef const void *GDExtensionConstObjectPtr;
+ typedef void *GDExtensionUninitializedObjectPtr;
  typedef void *GDExtensionTypePtr;
  typedef const void *GDExtensionConstTypePtr;
+ typedef void *GDExtensionUninitializedTypePtr;
  typedef const void *GDExtensionMethodBindPtr;
  typedef int64_t GDExtensionInt;
  typedef uint8_t GDExtensionBool;
```

The reason is that we don't treat those typedefs as `void*` but rather opaque pointers, so they have all different types. Since lots of signatures changed, I would need to update our counterparts for this.

touilleMan commented on core/extension/gdextension_interface.h:
```diff
@@ -139,16 +139,40 @@ typedef enum {
 
 } GDExtensionVariantOperator;
 
+// In this API there is multiple functions which expect the caller to pass a pointer
```
done :+1:

akien-mga commented:
Thanks!

saki7 commented:
To be honest, I was surprised that this PR was merged too quickly.

You can see in the comments above that the committer of Rust binding was showing direct concern about the changes. The author of this PR was asking about how the changes could be handled on Rust-side. The contributor of GDExtension was also kindly asking about the same issue on the contributors chat [(link)](https://chat.godotengine.org/channel/gdextension/thread/8MJYpSXgDjmTg8rZH).

I personally don't write Rust, I am just a godot-cpp user, but I believe that the Rust binding is an important part in the GDExtension ecosystem. I don't think that the PR should have been merged before receiving a reply from the participant.

akien-mga commented:
I interpreted @Bromeon's comment as highlighting that some work will be needed on the Rust bindings side, but not that anything is wrong per se with this PR.

The further reply from @touilleMan outlined that the way Rust handles this would be a good improvement to make for C too if possible, in a follow up PR. So I took this as a hint that this PR is fine to merge for now, Rust bindings will need to do some extra changes, and further work may be done to unify things further.

If I misinterpreted and merged too fast, I'm sorry. But there's also https://github.com/godotengine/godot/pull/76406 which was building upon this and waiting for a merge, and I keep hearing that GDExtension PRs don't get merged fast enough, so...

saki7 commented:
@akien-mga Thanks for spending your time investigating this issue. I truly appreciate that the PR was merged for now, and I have no further argument if @Bromeon has no concern about the future work required on Rust side.

> If I misinterpreted and merged too fast, I'm sorry. But there's also https://github.com/godotengine/godot/pull/76406 which was building upon this and waiting for a merge, and I keep hearing that GDExtension PRs don't get merged fast enough, so...

The issue regarding the merging process of GDExtension is a broader matter. It's not limited to this specific PR. I can't pretend not to be surprised if some PRs are going to be merged suddenly whereas other PRs which had received enough approvements are kept unhandled. Many PRs related to GDExtension needs more attension, and I believe the real problem is the lack of the maintainers.

Recently on #gdextension channel in Godot Contributors Chat, I have proposed to promote some contributor to maintainer. Doing so will fundamentally improve the situation, and the change will motivate people (like me) who are trying to submit new patches. I've got positive responses from several people (including the people from GDExtension team), but the actual movement has not happened yet.

aaronfranke commented:
@saki7 There is no need to do such a promotion. Anyone is able to, and welcome to, contribute to the engine. Anyone can review PRs. Regardless of whether or not the review is from a trusted member, additional effort towards PR reviewing is always appreciated. If someone continuously puts in effort, their efforts will be noticed.

saki7 commented:
@aaronfranke I know that's how OSS works. I'm afraid there's some misunderstanding, and the actual situation is worse than you think. Let me clarify the recent situation of GDExtension team:

- Many people, including both contributors and non-contributors are submitting PRs to GDExtension.
- However, there is currently no person in charge of the entire codebase.
   - You can confirm this by looking at the fact that no collaborator has the "star" mark on the [member list](https://godotengine.org/teams/).
   - Two people are known as the top-level contributors: mux213 and vnen. When I asked question on the Contributors Chat, mux213 answered [(link)](https://chat.godotengine.org/channel/gdextension?msg=izYq7YMhdgIJfbMQG):
     >  mux213: The problem is that we currently don’t have any such person for GDextension. vnen and myself come close i guess but neither of us has the bandwidth to become actively involved to properly review the work and so it stalls. 
  - mux213 voted dsnopek to take the role.
- Since few months ago, @dsnopek had been organizing GDExtension team meetings multiple times.
  - The attendee consists of top-level contributors in the GDExtension domain.
  - But even when the PR was confirmed on the meeting, it was usual that nobody actually presses the "Merge" button.
  - Even the PRs submitted by GDExtension team member have been stalling for days and weeks.

This is why I'm repeatedly asking for promotion. GDExtension team is also aware of the situation, and people are trying to advance.

You could check the log for [`#gdextension` channel](https://chat.godotengine.org/channel/gdextension) on the Godot Contributors Chat. The relevant conversation mostly happened after May 11, 2023.

This is getting off-topic, so please don't reply directly on this PR. If you need to ask me something, I'm also available on the Contributors Chat.
